dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+behaviour as in for string in Python if the operator exists in Python. For
+those operators which are not defined in Python, the result should be False.
+
+User supplied constants are always encoded as strings with either ``'`` or
+``"`` quote marks. Note that backslash escapes are not defined, but existing
+implementations do support them them. They are not included in this
+specification because none of the variables to which constants can be compared
+contain quote-requiring values.
+
+The variables in the marker grammar such as "os_name" resolve to values looked
+up in the Python runtime. With the exception of "extra" all values are defined
+on all Python versions today - it is an error in the implementation of markers
+if a value is not defined.
+
+Unknown variables must raise a SyntaxError rather than resulting in a
+comparison that evaluates to True or False.
+
+The "extra" variable is special. It is used by wheels to signal which
+specifications apply to a given extra in the wheel ``METADATA`` file, but
+since the ``METADATA`` file is based on a draft version of PEP-426, there is
+no current specification for this. Regardless, outside of a context where this
+special handling is taking place, the "extra" variable should result in a
+SyntaxError like all other unknown variables.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Marker
+     - Python equivalent
+     - Sample values
+   * - ``os_name``
+     - ``os.name``
+     - ``posix``, ``java``
+   * - ``sys_platform``
+     - ``sys.platform``
+     - ``linux``, ``darwin``, ``java1.8.0_51``
+   * - ``platform_machine``
+     - ``platform.machine()``
+     - ``x86_64``
+   * - ``python_implementation``
+     - ``platform.python_implementation()``
+     - ``CPython``, ``Jython``
+   * - ``platform_version``
+     - ``platform.version()``
+     - ``#1 SMP Fri Apr 25 13:07:35 EDT 2014``
+
+       ``Java HotSpot(TM) 64-Bit Server VM, 25.51-b03, Oracle Corporation``
+
+       ``Darwin Kernel Version 14.5.0: Wed Jul 29 02:18:53 PDT 2015; root:xnu-2782.40.9~2/RELEASE_X86_64``
+   * - ``python_version``
+     - ``platform.python_version()[:3]``
+     - ``3.4``, ``2.7``
+   * - ``python_full_version``
+     - see definition below
+     - ``3.4.0``, ``3.5.0b1``
+   * - ``extra``
+     - A SyntaxError except when defined by the context interpreting the
+       specification.
+     - ``test``
+
+Backwards Compatibility
+=======================
+
+Most of this PEP is already widely deployed and thus offers no compatibiltiy
+concerns.
+
+There are however a few points where the PEP differs from the deployed base.
+
+Firstly, PEP-440 direct references haven't actually been deployed in the wild,
+but they were designed to be compatibly added, and there are no known
+obstacles to adding them to pip or other tools that consume the existing
+dependency metadata in distributions - particularly since they won't be
+permitted to be present in PyPI uploaded distributions anyway.
```

Comment:
I remain unconvinced that we cannot, at the very least, define the possible dependencies up front. I have yet to hear a use case where people are expecting a package to have completely arbitrary dependencies. The ones I've heard seem to fall into two categories:

* We don't want to hard require certain things, but depending on features or what is available we might select different dependencies.
* We have a broad range for a dependency, but once we've built the binary then we need to restrict that to versions that will be ABI compatible with what we've built against.

Neither one of those are situations where the package authors can't enumerate the entire list of possible dependencies before hand, and then when the wheel gets built it can dynamically restrict the total items that it depends on from that total list of possibilities.

We don't really want to special case PyPI here, we've gone to some lengths to try and remove special casing PyPI other than the fact it's the default repository and we define things in terms of whether it's a public repository or a private repository.

That's neither here nor there though because this PEP can (and should) further build on the direct dependency syntax introduced in PEP 440. We just won't allow that particular part of the PEP in pip until we have a solution to being able to prevent people using them on PyPI (one way or another).